Challenge-4/Bot_1/src/helper.ts

```diff
@@ -1,67 +1,73 @@
-  const key = `${assetToken}-${blockNumber}`;
-  if (cache.has(key)) {
-    return cache.get(key) as { [name: string]: string };
+  constructor(provider: ethers.providers.Provider) {
+    this.provider = provider;
+    this.cache = new LRUCache<string, { [name: string]: string }>({
+      max: 1000,
+    });
   }
-  let infoArray: { [name: string]: string } = {};
-  let infoObject: any;
-  let index = 0; // Start from 0, increment before fetching next
 
-  while (true) {
-    try {
-      infoObject = await assetInfo.getAssetInfo(index, {
-        blockTag: blockNumber,
-      });
-
-      const collateralName =
-        "Collateral Asset - " +
-        (await getCollateralName(provider, infoObject.asset, blockNumber));
-      infoArray[collateralName] = infoObject.asset;
-      index++; // Increment after successful fetch
-    } catch (error) {
-      break; // Exit loop on error
+  public async getCollateralAssets(
+    assetToken: string,
+    abi: string[],
+    blockNumber: number,
+  ): Promise<{ [name: string]: string }> {
+    const assetContract = new ethers.Contract(assetToken, abi, this.provider);
+    const key = `${assetToken}-${blockNumber}`;
+    if (this.cache.has(key)) {
+      return this.cache.get(key) as { [name: string]: string };
     }
-  }
+    let collateralAssets: { [name: string]: string } = {};
+    let index = 0; // Start from 0, increment before fetching next
 
-  cache.set(key, infoArray);
-  return infoArray;
-}
+    while (true) {
+      try {
+    let collateralAssetInfo: any;
 
-// Fetching names of the collteral assets for specifying in metadata
-async function getCollateralName(
-  provider: ethers.providers.Provider,
-  tokenAddress: string,
-  blockNumber: number,
-): Promise<string> {
-  const erc20Abi = ["function name() view returns (string)"];
-  const tokenContract = new ethers.Contract(tokenAddress, erc20Abi, provider);
+        collateralAssetInfo = await assetContract.getAssetInfo(index, {
+          blockTag: blockNumber,
+        });
 
-  const tokenName = await tokenContract.name({ blockTag: blockNumber });
-  return tokenName;
-}
+        const collateralName =
+          "Collateral Asset - " +
+          (await this.getCollateralName(collateralAssetInfo.asset, blockNumber));
+          collateralAssets[collateralName] = collateralAssetInfo.asset;
+        index++; // Increment after successful fetch
+      } catch (error) {
+        break; // Exit loop on error
+      }
+    }
 
-export const getAddress = async (networkManager: any) => {
-  return networkManager.get("usdc");
-};
+    this.cache.set(key, collateralAssets);
+    return collateralAssets;
+  }
 
-export const getConfigurator = async (chainId: number) => {
-  if (chainId == 1) {
-    return CONFIGURATOR_PROXY;
-  } else if (chainId == 42161) {
-    return CONFIGURATOR_PROXY_ARB;
+  // Fetching names of the collteral assets for specifying in metadata
+  private async getCollateralName(
```

Comment:
This function should have caching, since right now you are making a call using an endpoint to fetch the `name`. Since it should not change, caching is a quick way to save unnecessary network calls. If this were a paid endpoint from something like Alchemy, that means money saved. 😎 

And the `key` used for caching shouldn't have the block number, since the name won't change. When I say `key`, I mean like the one from above `const key = '${assetToken}-${blockNumber}';`, you wouldn't need the `blockNumber`.